snappass/main.py

```diff
@@ -136,51 +138,71 @@ def empty(value):
         return True
 
 
-def clean_input():
+def clean_input(password=None, ttl=None):
     """
     Make sure we're not getting bad data from the front end,
     format data to be machine readable
     """
-    if empty(request.form.get('password', '')):
+    if empty(password):
         abort(400)
-
-    if empty(request.form.get('ttl', '')):
+    if empty(ttl):
         abort(400)
-
-    time_period = request.form['ttl'].lower()
+    time_period = ttl.lower()
     if time_period not in TIME_CONVERSION:
         abort(400)
+    return ttl,
 
-    return TIME_CONVERSION[time_period], request.form['password']
-
-
-@app.route('/', methods=['GET'])
-def index():
-    return render_template('set_password.html')
-
-
-@app.route('/', methods=['POST'])
-def handle_password():
-    ttl, password = clean_input()
-    token = set_password(password, ttl)
 
+def set_base_url(req):
     if NO_SSL:
         if HOST_OVERRIDE:
             base_url = f'http://{HOST_OVERRIDE}/'
         else:
-            base_url = request.url_root
+            base_url = req.url_root
     else:
         if HOST_OVERRIDE:
             base_url = f'https://{HOST_OVERRIDE}/'
         else:
-            base_url = request.url_root.replace("http://", "https://")
+            base_url = req.url_root.replace("http://", "https://")
     if URL_PREFIX:
         base_url = base_url + URL_PREFIX.strip("/") + "/"
-    link = base_url + quote_plus(token)
-    if request.accept_mimetypes.accept_json and not request.accept_mimetypes.accept_html:
+    return base_url
+
+
+@app.route('/', methods=['GET'])
+def index():
+    return render_template('set_password.html')
+
+
+@app.route('/', methods=['POST'])
+def handle_password():
+    password = request.form.get('password')
+    ttl = request.form.get('ttl')
+    if clean_input(password, ttl):
+        ttl = TIME_CONVERSION[ttl.lower()]
+        token = set_password(password, ttl)
+        base_url = set_base_url(request)
+        link = base_url + quote_plus(token)
+        if request.accept_mimetypes.accept_json and not \
+           request.accept_mimetypes.accept_html:
+            return jsonify(link=link, ttl=ttl)
+        else:
+            return render_template('confirm.html', password_link=link)
+    else:
+        abort(500)
+
+
+@app.route('/api/set_password/<string:password>', methods=['POST'])
+@app.route('/api/set_password/<string:password>/<string:ttl>', methods=['POST'])
+def api_handle_password(password: str, ttl: str = 'two weeks'):
```

Comment:
We should pass in the password in post payload instead of the path parameter, which could be accidentally exposed in plaintext in middleware or reverse proxy logs

Reply:
`ttl` can be an optional key in the body. It would be better to take an integer `ttl` <=1209600 to support max flexibility.

Reply:
See the latest commit